src/modulus_convert.rs

```diff
@@ -0,0 +1,148 @@
+use crate::error::BoxError;
+use crate::field::Field;
+use crate::helpers::ring::Ring;
+use crate::prss::Participant;
+use crate::replicated_secret_sharing::ReplicatedSecretSharing;
+use crate::securemul::{ProtocolContext, SecureMul};
+use bit_vec::BitVec;
+
+pub enum HelperIdentity {
+    H1,
+    H2,
+    H3,
+}
+
+pub struct RandomShareGenerationHelper<'a> {
+    rng: &'a Participant,
+    identity: HelperIdentity,
+}
+
+impl<'a> RandomShareGenerationHelper<'a> {
+    #[must_use]
+    pub fn new(rng: &'a Participant, identity: HelperIdentity) -> Self {
+        Self { rng, identity }
+    }
+
+    #[must_use]
+    pub fn split_binary<T: Field>(
+        &self,
+        left: bool,
+        right: bool,
+    ) -> (
+        ReplicatedSecretSharing<T>,
+        ReplicatedSecretSharing<T>,
+        ReplicatedSecretSharing<T>,
+    ) {
+        let left = u128::from(left);
+        let right = u128::from(right);
+        match self.identity {
+            HelperIdentity::H1 => (
+                ReplicatedSecretSharing::new(T::from(left), T::ZERO),
+                ReplicatedSecretSharing::new(T::ZERO, T::from(right)),
+                ReplicatedSecretSharing::new(T::ZERO, T::ZERO),
+            ),
+            HelperIdentity::H2 => (
+                ReplicatedSecretSharing::new(T::ZERO, T::ZERO),
+                ReplicatedSecretSharing::new(T::from(left), T::ZERO),
+                ReplicatedSecretSharing::new(T::ZERO, T::from(right)),
+            ),
+            HelperIdentity::H3 => (
+                ReplicatedSecretSharing::new(T::ZERO, T::from(right)),
+                ReplicatedSecretSharing::new(T::ZERO, T::ZERO),
+                ReplicatedSecretSharing::new(T::from(left), T::ZERO),
+            ),
+        }
+    }
+
+    /// Runs the protocol to generate a pair of secret sharings of a random value "r" in [0, 1]
+    /// None of the three helpers will know what the value of "r" is
+    ///
+    /// ## Errors
+    /// Lots of things may go wrong here, as this method calls the "Secure Mult" implementation
+    /// If there is a timeout (or other error) it should be signalled in the response
+    pub async fn gen_r_pairs<T: Field, R: Ring>(
+        &self,
+        idx: u128,
+        ctx: &ProtocolContext<'_, R>,
+    ) -> Result<Vec<(bool, ReplicatedSecretSharing<T>)>, BoxError> {
+        let mut idx = idx;
+
+        let (left, right) = self.rng.generate_values(idx);
```

Comment:
looking at examples in `Modulus conversion` doc and the documentation to `Participant` struct I am not sure I fully understand this. In the doc, each helper wants to generate (Xi, Xi+1) share: H1 -> (x1, x2), H2 -> (x2, x3) and H3 -> (x3, x1) using helper's own seed and seed of the helper on the right side.

I could be misreading this, but `Participant` struct has a note about seeds being provided for left and right parties for each helper. 

```
    /// Generate an additive share of zero.
    /// Each party generates two values, one that is shared with the party to their left,
    /// one with the party to their right.
```

Am I reading this correctly?

Reply:
So this line generates 2 random u128 values, the first is the same random number as the helper to the left will generate (for this index) and the second is the same random number as the helper to the right will generate (for this index).

Next, we will interpret these u128s as a sequence of just 128 bits. So now we have 128 binary secret sharings of the form:

H1 -> (x1, x2), H2 -> (x2, x3), H3 -> (x3, x1)

So each of these is a secret sharing in Z_2, which is either 0 or 1. No helper knows all 3 shares, so none of them know if it's a secret sharing of 0 or 1. Each one just has two bits.

The rest of this protocol is there to generate a *matching* secret sharing of the same value (either 0 or 1), but in a different modulus (i.e. Fp31).

Reply:
Put // in front of that and add it to the code?